sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala

```diff
@@ -918,6 +918,297 @@ class RocksDBStateStoreSuite extends StateStoreSuiteBase[RocksDBStateStoreProvid
+      assert(iterator(store, colFamily2).isEmpty)
+      put(store, ("a", 1), 1, colFamily2)
+      assert(valueRowToData(get(store, "a", 1, colFamily2)) === 1)
+
+      store.commit()
+
+      // reload version 0
+      store = provider.getStore(0)
+      assert(get(store, "a", 1, colFamily1) === null)
+      assert(iterator(store, colFamily1).isEmpty)
+
+      store = provider.getStore(1)
+      // version 1 data recovered correctly
+      assert(valueRowToData(get(store, "a", 1, colFamily1)) == 1)
+      assert(iterator(store, colFamily1).toSet === Set((("a", 1), 1)))
+      // make changes but do not commit version 2
+      put(store, ("b", 1), 2, colFamily1)
+      assert(valueRowToData(get(store, "b", 1, colFamily1)) === 2)
+      assert(iterator(store, colFamily1).toSet === Set((("a", 1), 1), (("b", 1), 2)))
+      // version 1 data recovered correctly
+      assert(valueRowToData(get(store, "a", 1, colFamily2))== 1)
+      assert(iterator(store, colFamily2).toSet === Set((("a", 1), 1)))
+      // make changes but do not commit version 2
+      put(store, ("b", 1), 2, colFamily2)
+      assert(valueRowToData(get(store, "b", 1, colFamily2))=== 2)
+      assert(iterator(store, colFamily2).toSet === Set((("a", 1), 1), (("b", 1), 2)))
+
+      store.commit()
+    }
+  }
+
+  Seq(
+    NoPrefixKeyStateEncoderSpec(keySchema), PrefixKeyScanStateEncoderSpec(keySchema, 1)
+  ).foreach { keyEncoder =>
+    testWithColumnFamilies(s"validate rocksdb " +
+      s"${keyEncoder.getClass.toString.split('.').last} correctness",
+      TestWithBothChangelogCheckpointingEnabledAndDisabled) { colFamiliesEnabled =>
+        tryWithProviderResource(newStoreProvider(keySchema, keyEncoder,
+          colFamiliesEnabled)) { provider =>
+          val store = provider.getStore(0)
+
+          val cfName = if (colFamiliesEnabled) "testColFamily" else "default"
+          if (colFamiliesEnabled) {
+            store.createColFamilyIfAbsent(cfName, keySchema, valueSchema, keyEncoder)
+          }
+
+          var timerTimestamps = Seq(931L, 8000L, 452300L, 4200L, -1L, 90L, 1L, 2L, 8L,
+            -230L, -14569L, -92L, -7434253L, 35L, 6L, 9L, -323L, 5L)
+          // put & get, iterator
+          timerTimestamps.foreach { ts =>
+            val keyRow = if (ts < 0) {
+              dataToKeyRow("a", ts.toInt)
+            } else dataToKeyRow(ts.toString, ts.toInt)
+            val valueRow = dataToValueRow(1)
+            store.put(keyRow, valueRow, cfName)
+            assert(valueRowToData(store.get(keyRow, cfName)) === 1)
+          }
+          assert(store.iterator(cfName).toSeq.length == timerTimestamps.length)
+
+          // remove
+          store.remove(dataToKeyRow(1L.toString, 1.toInt), cfName)
+          timerTimestamps = timerTimestamps.filter(_ != 1L)
+          assert(store.iterator(cfName).toSeq.length == timerTimestamps.length)
+
+          // prefix scan
+          if (!keyEncoder.getClass.toString.contains("No")) {
```

Comment:
nit: Probably clearer to do pattern matching despite it is a couple lines longer. Or one-two liner with isInstanceOf.